crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,394 @@
+    }
+
+    /// Unions all checkpoint accumulators at the end of the epoch to generate the
+    /// root state hash and saves it. This function is guaranteed to be idempotent (despite the
+    /// underlying data structure not being) as long as it is not called in a multi-threaded
+    /// context.
+    async fn accumulate(&self, last_state_of_epoch: State) -> Result<(), TypedStoreError> {
+        let State {
+            epoch,
+            checkpoint_seq_num: last_checkpoint,
+            ..
+        } = last_state_of_epoch;
+
+        if self
+            .store
+            .tables
+            .root_state_hash_by_epoch
+            .contains_key(&epoch)?
+        {
+            return Ok(());
+        }
+
+        let (next_to_accumulate, mut root_state_hash) = self
+            .store
+            .tables
+            .root_state_hash_by_checkpoint
+            .iter()
+            .skip_to_last()
+            .next()
+            .map(|(highest, hash)| (highest.saturating_add(1), hash))
+            .unwrap_or((0, Accumulator::default()));
+
+        for i in next_to_accumulate..=last_checkpoint {
+            let acc = self
+                .store
+                .tables
+                .state_hash_by_checkpoint
+                .get(&i)?
+                .unwrap_or_else(|| {
+                    panic!("Accumulator for checkpoint sequence number {i:?} not present in store")
+                });
+            root_state_hash.union(&acc);
+        }
+
+        // We want to enforce that this table is append only, otherwise we
+        // may end up re-accumulating the same state
+        assert!(
+            !self
+                .store
+                .tables
+                .root_state_hash_by_checkpoint
+                .contains_key(&last_checkpoint)?,
+            "StateAccumulator: root state hash already exists for checkpoint {last_checkpoint:?}"
+        );
+
+        // Update root_state_hash tables atomically
+        let batch = self.store.tables.root_state_hash_by_checkpoint.batch();
+
+        let batch = batch.insert_batch(
+            &self.store.tables.root_state_hash_by_checkpoint,
+            std::iter::once((&last_checkpoint, root_state_hash.clone())),
+        )?;
+        let batch = batch.insert_batch(
+            &self.store.tables.root_state_hash_by_epoch,
+            std::iter::once((&epoch, root_state_hash.clone())),
+        )?;
+
+        batch.write()?;
+        self.store
+            .root_state_notify_read
+            .notify(&epoch, &root_state_hash);
+
+        Ok(())
+    }
+}
+
+fn hash_checkpoint_effects(
+    state: State,
+    store: Arc<StateAccumulatorStore>,
+) -> SuiResult<Accumulator> {
+    let seq_num = state.checkpoint_seq_num;
+
+    if store
+        .tables
+        .state_hash_by_checkpoint
+        .contains_key(&seq_num)
+        .unwrap()
+    {
+        return Err(SuiError::from(
+            format!("StateAccumulator: checkpoint already hashed: {seq_num:?}").as_str(),
```

Comment:
we could just return the previously computed accumulator here instead of an error